ompi/mca/coll/ucc/coll_ucc.h

```diff
@@ -0,0 +1,107 @@
+/**
+  Copyright (c) 2021      Mellanox Technologies. All rights reserved.
+  $COPYRIGHT$
+
+  Additional copyrights may follow
+
+  $HEADER$
+ */
+
+#ifndef MCA_COLL_UCC_H
+#define MCA_COLL_UCC_H
+
+#include "ompi_config.h"
+#include "mpi.h"
+#include "ompi/mca/mca.h"
+#include "opal/memoryhooks/memory.h"
+#include "opal/mca/memory/base/base.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/communicator/communicator.h"
+#include "ompi/attribute/attribute.h"
+#include "ompi/op/op.h"
+#include "coll_ucc_debug.h"
+#include <ucc/api/ucc.h>
+
+BEGIN_C_DECLS
+
+#define COLL_UCC_CTS (UCC_COLL_TYPE_BARRIER   | UCC_COLL_TYPE_BCAST    | \
+                      UCC_COLL_TYPE_ALLREDUCE | UCC_COLL_TYPE_ALLTOALL | \
+                      UCC_COLL_TYPE_ALLTOALLV)
+
+#define COLL_UCC_CTS_STR "barrier,bcast,allreduce,alltoall,alltoallv"
+
+struct mca_coll_ucc_component_t {
+    mca_coll_base_component_2_4_0_t super;
+    int                             ucc_priority;
+    int                             ucc_verbose;
+    int                             ucc_enable;
+    int                             ucc_np;
+    char                           *cls;
+    char                           *cts;
+    const char                     *compiletime_version;
+    const char                     *runtime_version;
+    bool                            libucc_initialized;
+    ucc_lib_h                       ucc_lib;
+    ucc_lib_attr_t                  ucc_lib_attr;
+    ucc_coll_type_t                 cts_requested;
+    ucc_context_h                   ucc_context;
+};
+typedef struct mca_coll_ucc_component_t mca_coll_ucc_component_t;
+
+OMPI_MODULE_DECLSPEC extern mca_coll_ucc_component_t mca_coll_ucc_component;
+
+/**
+ * UCC enabled communicator
+ */
+struct mca_coll_ucc_module_t {
+    mca_coll_base_module_t              super;
+    ompi_communicator_t*                comm;
+    int                                 rank;
+    ucc_team_h                          ucc_team;
+    mca_coll_base_module_allreduce_fn_t previous_allreduce;
+    mca_coll_base_module_t*             previous_allreduce_module;
+    mca_coll_base_module_barrier_fn_t   previous_barrier;
+    mca_coll_base_module_t*             previous_barrier_module;
+    mca_coll_base_module_bcast_fn_t     previous_bcast;
+    mca_coll_base_module_t*             previous_bcast_module;
+    mca_coll_base_module_alltoall_fn_t  previous_alltoall;
+    mca_coll_base_module_t*             previous_alltoall_module;
+    mca_coll_base_module_alltoallv_fn_t previous_alltoallv;
+    mca_coll_base_module_t*             previous_alltoallv_module;
```

Comment:
how about non-blocking collectives?

Reply:
updated the PR to support NB collectives using UCC API completion callback. plz review.